File: rotate-image.cc
// Rotate Image
#define REP(i, n) for (int i = 0; i < (n); i++)

class Solution {
public:
  void rotate(vector<vector<int> > &a) {
    int n = a.size();
    REP(i, n-1)
      REP(j, n-1-i)
        swap(a[i][j], a[n-1-j][n-1-i]);
    REP(i, n/2)
      swap_ranges(a[i].begin(), a[i].end(), a[n-1-i].begin());
  }
};

File: rotate-image.rs
impl Solution {
    pub fn rotate(a: &mut Vec<Vec<i32>>) {
        a.reverse();
        let n = a.len();
        for i in 1..a.len() {
            let (before, after) = a.split_at_mut(i);
            for j in 0..i {
                std::mem::swap(&mut after[0][j], &mut before[j][i]);
            }
        }
    }
}
